Use libmkl_rt, or statically link against MKL

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31162 Don't use RTLD_GLOBAL to load _C.
* #31165 Pick parallel MKL implementation over sequential implementation.
* **#31177 Use libmkl_rt, or statically link against MKL**
* #31161 Actually get rid of Windows hacks in cpp_extensions.
* #31160 Stop using ctypes to interface with CUDA libraries.
* #31176 Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line.
* #31157 Add missing TORCH_CUDA_API annotation to throw_nccl_error
* #31155 Remove dead CAFFE2_LIBS variable
* #31152 Remove LibIRC logic from cmake.

Part of removing RTLD_GLOBAL for _C.

MKL doesn't like being in a local scope, unless you use the libmkl_rt.so
variant, so I adapted our CMake configuration to either accept that or
only allow static linking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Use libmkl_rt, or statically link against MKL

Part of removing RTLD_GLOBAL for _C.

MKL doesn't like being in a local scope, unless you use the libmkl_rt.so
variant, so I adapted our CMake configuration to either accept that or
only allow static linking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 2747d2b:
* **1/1** broken upstream at merge base 2488231 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=2488231fe32e6e2c86f58acaa34766a6daa19235&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2747d2b026e42a3c17a04beb4cfabd9f29a14955).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_xla_linux_xenial_py3_6_clang7_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Use libmkl_rt, or statically link against MKL"


Part of removing RTLD_GLOBAL for _C.

MKL doesn't like being in a local scope, unless you use the libmkl_rt.so
variant, so I adapted our CMake configuration to either accept that or
only allow static linking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
